### PR TITLE
Move RemoveOrphanFilesAction run in Driver to Excutor

### PR DESCRIPTION
**Pull request by zhangdove**
My use case :
1. table.expireSnapshots().cleanExpiredFiles(false).commit() [https://github.com/apache/iceberg/pull/1244]
2. actions.removeOrphanFiles().olderThan(t1).execute()

The first step takes about two seconds.However,in the second step of deleting the files, as the number of files increases, the deletion time becomes slower and slower, which is not what I want. If I do not understand the error, delete the file executed by a single thread in Spark Driver. Can we move the execution-deletion file from the Driver side to Spark's Executor to do multithreaded erasure of orphaned files?

### CODE REVIEW

File: spark/src/test/java/org/apache/iceberg/actions/TestRemoveOrphanFilesAction.java [hunk 1]
```diff
@@ -569,4 +569,39 @@ public void testRemoveOrphanFilesWithHadoopCatalog() throws InterruptedException
...
+
+    List<String> result = actions.removeOrphanFiles()
+        .olderThan(System.currentTimeMillis())
+        .executorParallelNum(10)
+        .execute();
+    Assert.assertEquals("Should delete only 20 files", 20, result.size());
+
+    // gets the number of tasks completed by the last stage
+    int completedTasks = spark.sparkContext().statusTracker()
+        .getStageInfo(23).get().numCompletedTasks();
```
Comment (zhangdove) [inline spark/src/test/java/org/apache/iceberg/actions/TestRemoveOrphanFilesAction.java:603]: Is there a better way to get the last stage ID?

File: spark/src/main/java/org/apache/iceberg/actions/RemoveOrphanFilesAction.java [hunk 2]
```diff
@@ -161,15 +172,44 @@ public RemoveOrphanFilesAction deleteWith(Consumer<String> newDeleteFunc) {
...
-        .onFailure((file, exc) -> LOG.warn("Failed to delete file: {}", file, exc))
-        .run(deleteFunc::accept);
+    paralleExecutor(sparkContext, orphanFiles, excutorTaskNum, deleteFunc);
 
     return orphanFiles;
   }
 
+  private void paralleExecutor(JavaSparkContext javaSc, List<String> orphanFiles,
+                              int numSlices, Consumer deleteConsumer) {
+    if (numSlices == 0) {
```
Comment (RussellSpitzer) [inline spark/src/main/java/org/apache/iceberg/actions/RemoveOrphanFilesAction.java:182]: I'm not a big fan of parameters which trigger different behaviors on certain values. Perhaps there should be a boolean which enables the different execution mode?

File: spark/src/main/java/org/apache/iceberg/actions/RemoveOrphanFilesAction.java [hunk 3]
```diff
@@ -161,15 +172,44 @@ public RemoveOrphanFilesAction deleteWith(Consumer<String> newDeleteFunc) {
...
+      Tasks.foreach(orphanFiles)
+          .noRetry()
+          .suppressFailureWhenFinished()
```
Comment (RussellSpitzer) [inline spark/src/main/java/org/apache/iceberg/actions/RemoveOrphanFilesAction.java:185]: We could just make this parallel as well as in the RemoveSnapshots code, where the user can pass in a different execution pool.

File: spark/src/main/java/org/apache/iceberg/actions/RemoveOrphanFilesAction.java [hunk 4]
```diff
@@ -161,15 +172,44 @@ public RemoveOrphanFilesAction deleteWith(Consumer<String> newDeleteFunc) {
...
+        Tasks.foreach(orphanFileList)
+            .noRetry()
+            .suppressFailureWhenFinished()
+            .onFailure((file, exc) -> LOG.warn("Failed to delete file: {}", file, exc))
+            .run(deleteConsumer::accept);
+      });
+    }
+  }
+
+  public static class DeleteConsumer implements Consumer<String>, Serializable {
```
Comment (RussellSpitzer) [inline spark/src/main/java/org/apache/iceberg/actions/RemoveOrphanFilesAction.java:200]: I wonder if it makes sense to add a Serializable Consumer class as a utility. Then have this DeleteConsumer extend that.

File: spark/src/test/java/org/apache/iceberg/actions/TestRemoveOrphanFilesAction.java [hunk 5]
```diff
@@ -569,4 +569,39 @@ public void testRemoveOrphanFilesWithHadoopCatalog() throws InterruptedException
...
+
+    Assert.assertEquals("Should executor only 10 tasks", 10, completedTasks);
```
Comment (RussellSpitzer) [inline spark/src/test/java/org/apache/iceberg/actions/TestRemoveOrphanFilesAction.java:605]: Message isn't clear here, "Executor should complete only 10 tasks"?

File: spark/src/main/java/org/apache/iceberg/actions/RemoveOrphanFilesAction.java [hunk 6]
```diff
@@ -106,6 +106,7 @@ public void accept(String file) {
     this.table = table;
     this.ops = ((HasTableOperations) table).operations();
     this.location = table.location();
+    this.deleteFunc = new DeleteConsumer(table);
```
Comment (RussellSpitzer) [inline spark/src/main/java/org/apache/iceberg/actions/RemoveOrphanFilesAction.java:109]: We set up a bit of a trap here. A serializable function is required here when using the Parallel path but we don't type check that we have a serializable consumer for this field. So we can end up with a user passing a non-serializable function here and getting a runtime error. Not sure if we want to lock this function to only serializable delete functions but I think we should at least document this and warn users.

Another early check could be to attempt to cast the Lambda to serializable before performing the parallelization and providing an explanation of the failure.

File: spark/src/main/java/org/apache/iceberg/actions/RemoveOrphanFilesAction.java [hunk 7]
```diff
@@ -161,15 +172,44 @@ public RemoveOrphanFilesAction deleteWith(Consumer<String> newDeleteFunc) {
...
-        .noRetry()
-        .suppressFailureWhenFinished()
-        .onFailure((file, exc) -> LOG.warn("Failed to delete file: {}", file, exc))
-        .run(deleteFunc::accept);
+    paralleExecutor(sparkContext, orphanFiles, excutorTaskNum, deleteFunc);
 
     return orphanFiles;
   }
 
+  private void paralleExecutor(JavaSparkContext javaSc, List<String> orphanFiles,
```
Comment (RussellSpitzer) [inline spark/src/main/java/org/apache/iceberg/actions/RemoveOrphanFilesAction.java:180]: I'm sure others have different opinions, but I would take the branch out of this function, currently it reads as though on line 175 all deletes are parallelized and then you need to look into this function to see that it's not actually the case. So either a rename "maybeExecuteParallel" or move the if branch out.

File: spark/src/test/java/org/apache/iceberg/actions/TestRemoveOrphanFilesAction.java (hunk 1 above)
Comment (RussellSpitzer) [reply spark/src/test/java/org/apache/iceberg/actions/TestRemoveOrphanFilesAction.java:603]: If the test here is just to determine the number of partitions in the RDD it's probably safer to expose that as a package protected member of the class and call .partitions. The magic number here is going to be brittle. 

One fun little game you could also play here to count tasks that were run, would be to implement another "DeleteConsumer" that just uses an accumulator and just add's 1 on each forEachParititon.

**Comment by aokolnychyi** [comment]
I would like to make sure I get the use case correctly. Is the main purpose to use `RemoveOrphanFilesAction` to clean up expired files? If so, I think that's a misuse of the API. Instead, one should use `ExpireSnapshotsAction` that Russell is working on in PR #1264.

The most expensive part in `RemoveOrphanFilesAction` is listing. The bigger table the more pressure you put on the storage. The number of orphan files should be usually small so the delete should be quick. This action should be called maybe once a month or so. `ExpireSnapshotsAction`, on the other hand, is meant to be called regularly. It will have an executor service to parallelize deletes.

**Comment by zhangdove** [comment]
@RussellSpitzer  Thanks for your review that has benefited me a lot.

@aokolnychyi As you said, maybe I don't use it in a reasonable way.
It wasn't until I saw this PR that I thought of moving the file deletion method to this method, so that I later made some other attempts to move the file deletion from driver to Executor. Thank you for your Suggestions.

I'm going to close this PR, and I'm going to keep watching and using master branch code

**Pull request closed without merging**